### DISCUSSION TITLE
Gotenberg + Webpack = Pdf?

### DISCUSSION THREAD

**Discussion by craigh** [symfony/symfony#42544] [Q&A]
Does anyone know how to use [Gotenberg](https://gotenberg.dev) to render a symfony page (rendered template + webpack assets) into a pdf?

I have Gotenberg up and running and I using it to combine existing PDFs. But if I use it to attempt to convert twig-rendered html into a pdf I get a page with no styles - e.g. the css stylesheets are not being used. For reference ALL my assets are rendered with an absolute url. `$encoreEntrypointLookup->reset()` is also done before each new page.

The Gotenberg docs indicate that all assets must be manually injected - which I am not doing. I'm hoping this is not truly needed.

**Comment by S-Tian86** [comment]
Hi, I have never used that library, I have only used TCPDF, but from what you say it should be similar.

I at least what I did was to have a template with the reusable styles between all the documents that I generate, and then when calling the library I "concatenate" the styles with the body of the message, leaving a single html with the styles inside, so that the generator does not have to go to look for it anywhere, and for the images, I pass them as base 64.

I hope I can guide you, or maybe I am totally wrong.

Greetings.

```php
        $html_body = $this->reemplazarMarcadores($marcadores, $script);
        
        /* Se obtienen los html de las diferentes secciones que componen el PDF */
        $html_estilos = $this->render('consultas/pdf/estilos.html.twig')->getContent();
        $html_titulo = $this->render('consultas/pdf/cert_antiguedad_titulo.html.twig', array('personal'=>$personal))->getContent();
        $html_footer = '<table><tr><td style="width: 35%;"></td><td style="text-align: center; width: 30%;"><img src="data:image/png;base64,'.$this->get('app.parametros')->getImg_firma().'"></td></tr></table>';

        /* Obtiene el servicio de generación de PDF */
        $pdf = $this->get('app.pdf_manager.printer');
        /* Setea si se imprimirá tanto el header como el footer, además de 
         * agregar una página vacía, el título y el cuerpo */
        
        $pdf->setImprimirHeader(true)
                ->setImprimirFooter(false)
                ->agregarPaginaVacia()
                ->agregarHTMLCell('','','',32,$html_estilos.$html_titulo)
                ->agregarHTMLCell('','','',35,$html_estilos.$html_body)
                ->agregarHTMLCell('','','',230,$html_footer);
        
        /* Dependiendo del parámetro $descargar de la URL, el archivo se visualiza o se descarga */
        $destino = $descargar ? 'D' : 'I';
        $nombreArchivo = sprintf('Certificado Antiguedad %s.pdf', $personal->getRut());
        
        /* Se genera el pdf a través de $pdf->Output() con el nombre del archivo
         * y el destino (ver en pantalla/descargar) */
        return $pdf->Output($nombreArchivo,$destino);
```

**Reply by craigh** [reply]
I appreciate you taking the time to respond. Unfortunately, Webpack handles styles (and javascript) differently and so it is not so straightforward. Thanks again!

**Reply by fd6130** [reply]
Well, i use dompdf-bundle. And just remind that most of the pdf library doesn't support latest CSS technology.

**Reply by S-Tian86** [reply]
Oh, It's ok. At least I tried. Lol.

**Comment by romaricdrigon** [comment]
Not really a straight answer, but in general, I found that PDF generation is supra tricky. In the end, now I'm using as much as possible Chrome headless to generates PDFs (or images). You can drive it using Pupeeter (in JS) or directly with Symfony Panther (with a custom webdriver command), so it integrates quite nicely.

**Reply by craigh** [reply]
Gotenberg * is * headless chrome

**Reply by romaricdrigon** [reply]
I googled before but couldn't find the info on their website - are you sure ?
Because if so, it should see the exact same result than you in a browser. But maybe some security restriction, are you serving over a specific domain or HTTPs ?

**Reply by craigh** [reply]
![Screen Shot 2021-08-19 at 6 10 09 AM](https://user-images.githubusercontent.com/350048/130051097-eb30347b-68c9-41da-bf23-ac1fb4495385.png)

**Comment by Warxcell** [comment]
I guess if you post the code you use for PDF conversion will be useful.
